### DISCUSSION TITLE
feature request: unzip support Extract Here (Smart)

### DISCUSSION THREAD

**Discussion by ivysrono** [doublecmd/doublecmd#1136] [Suggestions]
It is a feature of [Bandisoft](https://www.bandisoft.com/bandizip/help/auto_dest/).

> By right-clicking on an archive file, you can see 'Extract Here (Smart)' on the menu, and this feature works as follows:
>     If there is only one file/folder in an archive, it will extract to the target folder.
>     If all files in an archive is bundled in a single folder, it will extract to the target folder.
>     In other cases, it will extract to the 'Archive-Name' folder.
> Archive files compressed in Unix are often bundled into a single folder.
> So, if these files are extracted in the Archive-Name folder, it is inconvenient because two folders of the same name will be created.
> In this case, using 'Extract Here (Smart)' is useful.

If an archive contains 100 files, I want to avoid that these 100 files are extracted in the middle of the other files of the current folder. But I don't want to always create a folder, because sometimes, the archive contains only one folder, which can have a name different from the one of the archive. And even in this case, I don't want to have a subfolder for nothing.

I also don't want to create a subfolder if the archive contains only one file, which can also have a different name than the one of the archive.

Currently, to avoid these issues, the only way is either to open the archive to see what is inside, close it, and then use the context menu of Explorer; or to always extract in a subfolder, and then move its content if there is only one file or folder.

**Comment by ivysrono** [comment]
@alexx2000  [PeaZip](https://github.com/peazip/PeaZip) support this feature also:
https://peazip.github.io/how-to-extract-rar-zip-archives.html

> Smart new folder prevents multiple unexpected items being extracted to destination ("tarbomb"-like effect): automatically extracts to new folder, if the new folder contains a single object (file or directory) content is copied outside and the extra folder is removed (removing extra folder layer and preserving original names of archived folders), if copying the content outside the new folder fails, the content is kept in the new folder

**Comment by ivysrono** [comment]
@alexx2000 [vvyoko/SmartZip](https://github.com/vvyoko/SmartZip) support this feature by enhance 7-Zip.

**Comment by alexx2000** [comment]
I'm already implemented this feature, but don't commit it yet. Don't know how to name this option in the extract dialog.

**Comment by ivysrono** [comment]
How to open this option in DC 1.0.9?

**Comment by alexx2000** [comment]
Smart extract is a DC 1.1.0 only feature.

**Comment by ivysrono** [comment]
Great in 1.1.0  Beta, however, how to set it as default?

**Comment by alexx2000** [comment]
No way.

**Reply by ivysrono** [reply]
I'd like to build DC for myself.
I should edit `src/filesources/multiarchive/umultiarchivecopyoutoperation.pas` and `src/filesources/wcxarchive/uwcxarchivecopyoutoperation.pas` and anything else?

**Reply by Skif-off** [reply]
Try to open `src/fextractdlg.lfm`, find `object cbInSeparateFolder: TCheckBox` and add
```pas
        Checked = True
        State = cbChecked
```
after `Caption = -and-so-on`.
Or use `State = cbGrayed` instead of `State = cbChecked`.

**Reply by ivysrono** [reply]
Thank you very much!

**Comment by ivysrono** [comment]
Maybe add one parameter just like `ExtractHere`?
 
 Or remember the last option?

**Comment by ags1234** [comment]
Do you want to fork this project, to be qble to implement different things?

**Reply by ivysrono** [reply]
I don't know Pascal or Lazarus either.

**Reply by Skif-off** [reply]
The compilation of the program is well automated, see [Wiki](https://github.com/doublecmd/doublecmd/wiki/Development) and use the "From command line" item.
A little higher I wrote how you can change the default state, if your decision

> I'd like to build DC for myself.

is still relevant to you.

**Reply by ivysrono** [reply]
BTW, how to build Installation packages like `doublecmd-1.1.1.x86_64-win64.msi` and auto get portable packages like `doublecmd-1.1.1.x86_64-win64.zip`?

**Reply by Skif-off** [reply]
See the "install" folder (`create_packages.bat`). Or just replace `doublecmd.exe` and `doublecmd.zdli` in original distribution.

**Reply by ivysrono** [reply]
Great!

Just one question: https://github.com/doublecmd/doublecmd/discussions/1245
